feat: adds reusable actions (#296)
OAI/Overlay-Specification: versions/1.2.0-dev.md

Reviewer: We might want to update the Overlay section in Definitions to make it a bit less specific? Right now it says

An Overlay is a JSON or YAML structure containing an ordered list of [Action Objects](#overlay-actions) that are to be applied to the target document.

Author: attempted an update via a8c17199d140f823c33cb700046c41428312fb60

Happy to take better suggestions.

Author: (reviewed on the call, thanks)

### Overlay

> An Overlay document is represented in JSON or YAML and contains actions that are applied in order to an [[OpenAPI]] description. Each action uses a target expressed as a [[RFC9535|RFC9535 JSONPath]] query expression to identify where changes apply. [Reusable actions](#reusable-action-object) may be used to reduce duplication through references.

Reviewer 2: ```suggestion
An Overlay document is represented in JSON or YAML and contains [actions](#overlay-actions) that are applied in order to an [[OpenAPI]] description. Each action uses a target expressed as a [[RFC9535|RFC9535 JSONPath]] query expression to identify where changes apply. [Reusable actions](#reusable-action-object) may be used to reduce duplication through references.
```

| <a name="overlay-version"></a>overlay | `string` | **REQUIRED**. This string MUST be the [version number](#versions) of the Overlay Specification that the Overlay document uses. The `overlay` field SHOULD be used by tooling to interpret the Overlay document. |
| <a name="overlay-info"></a>info | [Info Object](#info-object) | **REQUIRED**. Provides metadata about the Overlay. The metadata MAY be used by tooling as required. |
| <a name="overlay-extends"></a>extends | `string` | URI reference that identifies the target document (such as an [[OpenAPI]] document) this overlay applies to. |

> | <a name="overlay-actions"></a>actions | [[Action Object](#action-object) or [Reusable Action Reference Object](#reusable-action-reference-object)] | **REQUIRED** An ordered list of actions to be applied to the target document. The array MUST contain at least one value. |

Reviewer 3: In the OAS, we use an escaped pipe for "or" with field types.

```suggestion
| <a name="overlay-actions"></a>actions | [[Action Object](#action-object) \| [Reusable Action Reference Object](#reusable-action-reference-object)] | **REQUIRED** An ordered list of actions to be applied to the target document. The array MUST contain at least one value. |
```

> | <a name="overlay-components"></a>components | [Components Object](#components-object) | A set of components to reuse across the Overlay Document. Optional. |

Reviewer: The required fields are required, this one isn't but it just means it doesn't need the entry.
```suggestion
| <a name="overlay-components"></a>components | [Components Object](#components-object) | A set of components to reuse across the Overlay Document. |
```

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Components Object

The object provides a set of components to be reused across the Overlay document.

##### Fixed Fields

| Field Name | Type | Description |
| ---- | :----: | ---- |

> | <a name="component-actions"></a>actions | Map(`string`, [Reusable Action Object](#reusable-action-object)) | A key-value set of reusable actions to reference in the actions through a [Reusable Action Reference Object](#reusable-action-reference-object). Optional. |

Reviewer: Same comment about the Optional field, we use REQUIRED if it's not optional

Author: ```suggestion
| <a name="component-actions"></a>actions | Map(`string`, [Reusable Action Object](#reusable-action-object)) | A key-value set of reusable actions to reference in the actions through a [Reusable Action Reference Object](#reusable-action-reference-object). |
```

> > Note: the target field for components actions is Optional.

Reviewer 4: This line confused me. Maybe it is out of place? If it refers to the `target` field of Reusable Action Object, it probably belongs in the section that describes that object.

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Reusable Action Object

This object defines a reusable action in `components.actions`.
It holds action fields that can be referenced from `actions` by a Reusable Action Reference Object, which MAY override those fields.
It MAY also define `parameters` and `environmentVariables` for string replacement during reference resolution.

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |

> | <a name="reusable-action-parameters"></a>parameters | [[Reusable action parameter object](#reusable-action-parameter-object)] | A list of parameters to be used during  string literal replacement. Optional. |

Reviewer 2: ```suggestion
| <a name="reusable-action-parameters"></a>parameters | [[Reusable Action Parameter Object](#reusable-action-parameter-object)] | A list of parameters to be used during  string literal replacement. Optional. |
```

| <a name="reusable-action-fields"></a>fields | [Action Object](#action-object) | An [Action Object](#action-object) where no field is required, except that the `target` field MUST NOT be provided. The `target` MUST be supplied by the [Reusable Action Reference Object](#reusable-action-reference-object). |

This object MAY be extended with [Specification Extensions](#specification-extensions).

A reusable action is similar to an action but differs in important ways:

1. Action fields are nested inside the `fields` property, not defined as peer properties. The `fields` object may omit any action field, and MUST NOT include the `target` field; the `target` MUST be supplied by the [Reusable Action Reference Object](#reusable-action-reference-object).
1. A Reusable Action Object has to be referenced by a Reusable Action Reference Object to have any effect on the OpenAPI description.

> 1. During resolution, the `fields` of the Reusable Action Object are resolved first, then any peer action fields on the [Reusable Action Reference Object](#reusable-action-reference-object) override the resolved values. The final result MUST be a valid [Action Object](#action-object).

Reviewer 4: I think this is the wrong place to describe resolution. I think it belongs in the description of the Reusable Action Reference Object belowl
```suggestion
```

#### Reusable Action Reference Object

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |
| <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, where the fragment syntax is JSON Pointer with the pointer prefix restricted to `/components/actions/`. Component action keys in this pointer MUST be encoded according to [[RFC6901], Section 3](https://www.rfc-editor.org/rfc/rfc6901#section-3). |
| <a name="reusable-action-reference-target"></a>target | `string` | **REQUIRED** A JSONPath query expression, as described in [[RFC9535]], referencing the target objects in the source document. |

> | <a name="reusable-action-reference-action-fields"></a>Any other field defined in the [Action Object](#action-object) | mixed | Any other field defined in the [Action Object](#action-object) to be used as an override to the value resolved in the reusable action. |

Reviewer 4: I don't think "resolved" is needed here. I also removed it later.
```suggestion
| <a name="reusable-action-reference-action-fields"></a>Any other field defined in the [Action Object](#action-object) | mixed | Any other field defined in the [Action Object](#action-object) to be used as an override to the value in the reusable action. |
```

This object MAY be extended with [Specification Extensions](#specification-extensions).

> 

Reviewer 4: Here. I've also reworded this text a bit.
```suggestion

Fields of the Reusable Action Reference Object are combined with the `fields` of the referenced [Reusable Action Object](#reusable-action-object) to produce an effective [Action Object](#action-object).
Fields of the Reusable Action Reference Object override fields of the same name in the `fields` of the Reusable Action Object. The final result MUST be a valid Action Object.
```

This could probably be tightened up a bit but I think it's close.

| <a name="reusable-action-reference-target"></a>target | `string` | **REQUIRED** A JSONPath query expression, as described in [[RFC9535]], referencing the target objects in the source document. |
| <a name="reusable-action-reference-action-fields"></a>Any other field defined in the [Action Object](#action-object) | mixed | Any other field defined in the [Action Object](#action-object) to be used as an override to the value in the reusable action. |

This object MAY be extended with [Specification Extensions](#specification-extensions).

Fields of the Reusable Action Reference Object are combined with the `fields` of the referenced [Reusable Action Object](#reusable-action-object) to produce an effective [Action Object](#action-object).

> Fields of the Reusable Action Reference Object override fields of the same name in the `fields` of the Reusable Action Object. The final result MUST be a valid Action Object.

Reviewer 2: ```suggestion
Fields of the Reusable Action Reference Object override fields of the same name in the `fields` of the Reusable Action Object. The final result MUST be a valid Action Object.

```
Empty line needed before heading

| <a name="reusable-action-fields"></a>fields | [Action Object](#action-object) | An [Action Object](#action-object) where no field is required, except that the `target` field MUST NOT be provided. The `target` MUST be supplied by the [Reusable Action Reference Object](#reusable-action-reference-object). |

This object MAY be extended with [Specification Extensions](#specification-extensions).

A reusable action is similar to an action but differs in important ways:

1. Action fields are nested inside the `fields` property, not defined as peer properties. The `fields` object may omit any action field, and MUST NOT include the `target` field; the `target` MUST be supplied by the [Reusable Action Reference Object](#reusable-action-reference-object).
1. A Reusable Action Object has to be referenced by a Reusable Action Reference Object to have any effect on the OpenAPI description.

#### Reusable Action Reference Object

> 

Reviewer: Let's add a sentence or two about what a resusable action reference object is and what it's for before we describe its fields

Author: added in 3515bf7dfbae7afa1fae1f0de73060a423b30feb

Author: (reviewed on the call, thanks)

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |
| <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, where the fragment syntax is JSON Pointer with the pointer prefix restricted to `/components/actions/`. Component action keys in this pointer MUST be encoded according to [[RFC6901], Section 3](https://www.rfc-editor.org/rfc/rfc6901#section-3). |
| <a name="reusable-action-reference-target"></a>target | `string` | **REQUIRED** A JSONPath query expression, as described in [[RFC9535]], referencing the target objects in the source document. |
| <a name="reusable-action-reference-action-fields"></a>Any other field defined in the [Action Object](#action-object) | mixed | Any other field defined in the [Action Object](#action-object) to be used as an override to the value in the reusable action. |

This object MAY be extended with [Specification Extensions](#specification-extensions).

> Fields of the Reusable Action Reference Object are combined with the `fields` of the referenced [Reusable Action Object](#reusable-action-object) to produce an effective [Action Object](#action-object).

Reviewer: Can we make it clearer that they're not combined? Whatever is in the reusable action is now the value that is used. So if an action has $ref then everything else is ignored (if I understood this!), but this sounds more open to mixing things together.

Author: The intent I was trying to convey was the following:
- if a field is present only in the reusable action, let's use that value.
- if a field is present only in the reference object, let's use that value.
- if a field is present in BOTH the reference and the reusable action, let's use the reference object value. (overrides)

Happy to take in any suggestion of words that better convey this.

Author: Collecting notes from the meeting:
- Because of the simplification we've done, there's no point using update/copy/remove next to $ref, otherwise, what are you using $ref even for? This should be mutually exclusive. Therefore no override mechanism is needed for those keywords
- Description should be defineable at both the reusable action level, and the reference point, for ergonomics. Because of that we need to have an arbitrage when both are present, and here "the one in the reference wins", and we need to update the wording.

Author: updated in d85d2641620b429bba92a52c8e34a50fe2620f24

It references the reusable action by `$ref` and MAY provide action fields that override values from the referenced reusable action for that use.

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |
| <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, where the fragment syntax is JSON Pointer with the pointer prefix restricted to `/components/actions/`. Component action keys in this pointer MUST be encoded according to [[RFC6901], Section 3](https://www.rfc-editor.org/rfc/rfc6901#section-3). |
| <a name="reusable-action-reference-target"></a>target | `string` | **REQUIRED** A JSONPath query expression, as described in [[RFC9535]], referencing the target objects in the source document. |
| <a name="action-description"></a>description | `string` | A description of the action. [[CommonMark]] syntax MAY be used for rich text representation. |

This object MAY be extended with [Specification Extensions](#specification-extensions).

> In case a description is defined in `fields` of the referenced [Reusable Action Object](#reusable-action-object) and in the Reference Object, the description from the Reference Object MUST be used.

Reviewer: Sorry, I thought more about this since we last discussed it and I think I've changed my mind... The reusable action is more generic, and the action that refers to it is more specific. So the overlay action's description should be used if it has one, and we can fall back to the reusable action's description if it doesn't.

Author: If I'm understanding your comment correctly, it should already be possible with the current wording? 

When:
- the reusable action has a description, but not the reference, use the one in the reusable action
- the reusable action does not have a description, but the reference does, use the one from the reference. 
- both have a description, use the one from the reference. 

Is this not what you expect anymore? Is the current wording unclear with regards to the intent?

Reviewer: I've read this a couple more times and I think it's the language that is confusing (I know I'm supposed to be helping with this, not making it more confusing!). I think part of the problem here is that we are referring to something as a "reference" that can define a target or a description alongside, but actually I think we should be referring to an "action" as defined in the actions section , or a "reusable action" as defined in the components section.

Specifically to the description field which I think is the only thing we can define in either location now: if only one or other location has it, use that, but if they both do then the action (defined in the actions: section) should be used, not the more generic one from the reusable action (defined in components.actions).

Author: ```suggestion
In case a description is defined in `fields` of the referenced [Reusable Action Object](#reusable-action-object) and in the [Reusable Action Reference Object](#reusable-action-reference-object), the description from the Reusable Action Reference Object MUST be used.
```

Is this better?

And the reason why we have a Reusable Action Reference Object different from the Action Object is because of the difference in required fields between the two. @handrews and I didn't want to write a bunch of conditional requirements after the experience in OpenAPI parameters, headers, etc...

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Reusable Action Object

> This object defines a reusable action in `components.actions`.

Reviewer 2: ```suggestion
This object defines a reusable action in [`components.actions`](#component-actions).
```

> It holds action fields that can be referenced from `actions` by a Reusable Action Reference Object, which MAY override those fields.

Reviewer 2: ```suggestion
It can be referenced from [`actions`](#overlay-actions) by a [Reusable Action Reference Object](#reusable-action-reference-object).
```
The "override" part is described in the Reusable Action Reference Section.

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Action Template Object

> This object represents one or more changes to be applied to the target document at the locations defined by the target JSONPath expression and reused in one of more action references.

Reviewer 4: ```suggestion
This object is a template that, when instantiated by an Action Template Instantiation, represents one or more changes to be applied to the target document at the locations defined by the target JSONPath expression and reused in one of more action references.
```

A reusable action is similar to an action but differs in important ways:

1. It may omit any action field, in particular the `target` field, as these may be supplied by the Reusable Action Reference Object.
1. Any action field in the Reusable Action may employ string interpolation using variables specified in the `parameters` or `environmentVariables` fields.
1. A Reusable Action MUST be referenced by a Reusable Action Reference Object to have any effect on the OpenAPI description.

#### Reusable Action Parameter Object

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |
| <a name="reusable-action-parameter-name"></a>name | `string` | **REQUIRED** The name of the parameter, MUST match the key for the string literal replacement expression. |
| <a name="reusable-action-parameter-default>default | Any | A default value for the parameter to use when no value is provided by the reference. Optional. |

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Reusable Action Reference Object

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |

> | <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, and that the fragment syntax is JSON Pointer, with the pointer prefix restricted to `/components/actions/`. |

Author: from talking with @ralfhandl on the call, yes that's what that effectively means, we need to call it out.

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Reusable Action Object

This object defines a reusable action in `components.actions`.
It holds action fields that can be referenced from `actions` by a Reusable Action Reference Object, which MAY override those fields.
It MAY also define `parameters` and `environmentVariables` for string replacement during reference resolution.

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |

> | <a name="reusable-action-parameters"></a>parameters | [[Reusable Action Parameter Object](#reusable-action-parameter-object)] | A list of parameters to be used during  string literal replacement. Optional. |

Author: ```suggestion
| <a name="reusable-action-parameters"></a>parameters | [[Reusable Action Parameter Object](#reusable-action-parameter-object)] | A list of parameters to be used during  string literal replacement. |
```

> | <a name="reusable-action-environment-variables"></a>environmentVariables | [[Reusable Action Parameter Object](#reusable-action-parameter-object)] | A list of environment variables to be used during  string literal replacement. Optional. |

Author: ```suggestion
| <a name="reusable-action-environment-variables"></a>environmentVariables | [[Reusable Action Parameter Object](#reusable-action-parameter-object)] | A list of environment variables to be used during  string literal replacement. |
```

> | <a name="reusable-action-action-fields"></a>Any field defined in the [Action Object](#action-object) | mixed | The [string literal replacement syntax](#string-literal-replacement-syntax) MAY be used in string fields (`target`, `copy`) and in key or value nodes within the `update` object. It MUST NOT be used in scalar non-string fields (for example `remove`), as that would produce an invalid Overlay document according to the schema. Replacements MUST be evaluated as the reusable action reference is being resolved. |

Reviewer: Do we have this situation of "the same fields as over there" anywhere else in our specifications? I'd like to see the fields actually named here, even if they just defer their definitions to the other object

Author: Kind of. JSON Schema references do that to some extent. I can't think of another example though. But I think I like @handrews suggestion to move all of that into a sub node called "fields", we can then simply refer to the schema of Action object with the two caveats that no field is required and the string interpolation syntax is supported. Better encapsulation.

Author: Resolving this here in favour of a central prior discussion here https://github.com/OAI/Overlay-Specification/pull/296#discussion_r3053717182

> | <a name="reusable-action-action-fields"></a>Any field defined in the [Action Object](#action-object) | mixed | The [string literal replacement syntax](#string-literal-replacement-syntax) MAY ONLY be used in string fields (`target`, `copy`) and in key or value nodes within the `update` object. It MAY NOT be used in scalar non-string fields (for example `remove`), as that would produce an invalid Overlay document according to the schema. Replacements MUST be evaluated as the reusable action reference is being resolved. |

Reviewer 3: MAY ONLY and MAY NOT are not RFC2119 terms.  What you want is:

```suggestion
| <a name="reusable-action-action-fields"></a>Any field defined in the [Action Object](#action-object) | mixed | The [string literal replacement syntax](#string-literal-replacement-syntax) MAY be used in string fields (`target`, `copy`) and in key or value nodes within the `update` object. It MUST NOT be used in scalar non-string fields (for example `remove`), as that would produce an invalid Overlay document according to the schema. Replacements MUST be evaluated as the reusable action reference is being resolved. |
```

This object MAY be extended with [Specification Extensions](#specification-extensions).

A reusable action is similar to an action but differs in important ways:

> 1. It may omit any action field, in particular the `target` field, as these may be supplied by the [Reusable Action Reference Object](#reusable-action-reference-object).

Reviewer: I had thought it was not allowed to include the target. The target has to be with the action, but the action definition can be referred to as one of the reusable items from components. What's the use case for supporting action?

Author: On this instance specifically, I think you've missed a few discussions, either directly here in the PR history, or on the Overlays call. I don't want to re-litigate the whole thing that late in the process. But at a high level:
- We believe that consistency is key, is there really any good reason to allow every field here but target?
- We believe that string interpolation + target are a powerful combination for advanced scenarios, that's actually lead to a huge reduction in my team's context "similar" actions.
- We believe we can also enable simpler scenarios (people who don't want to mess around with string interpolation), by allowing an override at the reference level.

Hopefully that's enough context, let me know if you still have questions or concerns.

Reviewer: This change brings multiple features in at once, and I'm uneasy that we've missed the purpose of Overlays and leapt in a new "more is more" direction with this proposal which seems to bring a lot of complexity and also brings back many aspects that have previously been discarded as not a good fit - either for being too much change in a single leap, or just being out of scope for what Overlays is.

I'm supportive of actions being reusable to avoid repetition. I'm not in favour of configurable targets which would allow a matrix of difficult to diagnose or reason about operations. I've consistently expressed these views, I'm not sure it's fair to frame it as me having "missed" something.

Author: I think I may have misunderstood your concern earlier. Based on your [initial comment on the earlier proposal](https://github.com/OAI/Overlay-Specification/pull/238#issuecomment-3538455079) and from what I recall from our conversations, I had understood the main sticking point to be the matrix aspect. Essentially, I got a sense that "nobody wants to do *math* in overlays, it brings too much cognitive load". This is why I rewrote the thing entirely, to start with new design principles and avoid carrying this over.
 
I'll be honest — getting this feedback at this stage is frustrating, as the draft has been open for two months and this specific concern about `target` [first came up 8 days ago](https://github.com/OAI/Overlay-Specification/pull/296#discussion_r3069695425). I recognize async communication is lossy, and I may have missed implicit signals, but it does make it harder to iterate when fundamental concerns surface this late.
 
Setting that aside, I'd like to better understand your concern about allowing string interpolation in `target`. Is it primarily about security, readability, invalid JSONPath after resolution, or overall spec complexity? If you can point to a concrete failure mode or example, that would help me think about alternatives.
 
The reason I've been pushing for interpolation in `target`/`copy` is that it enables patterns like [the sidecar example](https://github.com/OAI/Overlay-Specification/pull/238#issuecomment-3512628113). Another example would be cases where you have a dozen endpoints each with a different schema and want to introduce a status monitor for each, without relying on generic matching. I'm happy to update examples/test cases if they are not illustrative enough.
 
I'd also like to understand how you'd see this applying to `copy`, since I think consistency between the two matters.
 
I'm hoping this helps us narrow the disagreement and find a workable direction. :)

Reviewer: I'm sorry I've been unable to schedule myself into the meetings but let me try to recap where I think we are:
 - we're bringing in a few different features here
 - one is reusable actions to avoid copy/pasting the same action for lots of target expressions
 - we're also introducing parameters for those actions so that as well as allowing copy/paste, it's possible to adjust

All this is good so far, it's a big amount of change and does change the nature of Overlays as a project  - but innovation is good and we have clear use cases and good examples that will help.

What I'm questioning here is the reusable _targets_. We've talked about interpolation in the target of an action object. I understood the reusable actions to be copy or updates with parameters, and parameters potentially being pulled into the description or target of the actual (not reusable) action. But seeing reusable targets, I had doubts and had assumed that further discussion was still welcome.

Author: Hey @lornajane 
Apologies. During the last Overlay meeting, I had committed to put together a concrete example of why I believe string interpolation support in targets is important. Health issues and other priorities came in the way of doing this.

[Here is a gist that illustrates the difference in overlay experience](https://gist.github.com/baywet/3caafd66c9b431b0941c0354a5c145db), please have a look, and let me know whether it addresses your doubts regarding the usefulness of this aspect.

Essentially my concern is to enable reusable actions "components", where users only need to "fill in the parameters" and we get a consistent, easy to maintain, low thinking required experience from the users of the reusable actions.
I realized this maybe tailored to larger organizations, where large overlay documents are in use and where the people building the reusable actions and the one "referencing them" are not necessarily the same person. But I still think this is a genuine use case.

> 1. String interpolation in a Reusable Action Object using variables specified in the `parameters` or `environmentVariables` fields is only allowed in string fields (`target`, `copy`) and in key or value nodes within the `update` object. It is not allowed in scalar non-string fields such as `remove`.

Reviewer: Does it make sense to have remove in a reusable action? If not I am not sure I would refer to it here.

Author: I think having the reusable action ONLY define remove: true probably does not bring a lot of value in itself. However, if the description field and/or the target fields are using string interpolation, then it becomes interesting. We have this scenario for example where we remove a bunch of "preview" enum values across the document. In that case, it's fairly easy and useful to have a reusable action with the 3 fields, 2 templated (description and target)

Reviewer: Got it. I don't think configurable targets is a feature that should be in scope for this change.

Author: (resolving this one in favour of [that one](https://github.com/OAI/Overlay-Specification/pull/296#discussion_r3107544054) to avoid having the discussion at multiple places)

1. A Reusable Action Object has to be referenced by a Reusable Action Reference Object to have any effect on the OpenAPI description.

#### Reusable Action Parameter Object

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |
| <a name="reusable-action-parameter-name"></a>name | `string` | **REQUIRED** The name of the parameter, MUST match the key for the string literal replacement expression. |

> | <a name="reusable-action-parameter-default>default | `string` | A default string value for the parameter to use when no value is provided by the reference. Optional. |

Author: ```suggestion
| <a name="reusable-action-parameter-default>default | `string` | A default string value for the parameter to use when no value is provided by the reference. |
```

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Reusable Action Reference Object

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |

> | <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, and that the fragment syntax is JSON Pointer, with the pointer prefix restricted to `/components/actions/`. Component action keys in this pointer MUST be encoded according to [[RFC6901], Section 3](https://www.rfc-editor.org/rfc/rfc6901#section-3). |

Reviewer 3: I'd suggest a slight wording tweak:

```suggestion
| <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, where that the fragment syntax is JSON Pointer with the pointer prefix restricted to `/components/actions/`. Component action keys in this pointer MUST be encoded according to [[RFC6901], Section 3](https://www.rfc-editor.org/rfc/rfc6901#section-3). |
```

> | <a name="reusable-action-reference-parameter-values></a>parameterValues | `Map(string, string)` | A set of string values to use for the reusable action parameters, the key MUST match the parameter name. Optional. |

Author: ```suggestion
| <a name="reusable-action-reference-parameter-values></a>parameterValues | `Map(string, string)` | A set of string values to use for the reusable action parameters, the key MUST match the parameter name. |
```

> | <a name="reusable-action-reference-action-fields"></a>Any field defined in the [Action Object](#action-object) | mixed | Any field defined in the [Action Object](#action-object) to be used as an override to the value resolved in the reusable action. The [string literal replacement syntax](#string-literal-replacement-syntax) MAY NOT be used for any of the fields. Optional. |

Author: ```suggestion
| <a name="reusable-action-reference-action-fields"></a>Any field defined in the [Action Object](#action-object) | mixed | Any field defined in the [Action Object](#action-object) to be used as an override to the value resolved in the reusable action. The [string literal replacement syntax](#string-literal-replacement-syntax) MAY NOT be used for any of the fields. |
```

> | <a name="reusable-action-reference-action-fields"></a>Any field defined in the [Action Object](#action-object) | mixed | Any field defined in the [Action Object](#action-object) to be used as an override to the value resolved in the reusable action. The [string literal replacement syntax](#string-literal-replacement-syntax) MAY NOT be used for any of the fields. Optional. |

Reviewer 3: This seems to be missing the field name and type?  I see this is true in the other Object where I commented on RFC2119 usage as well?  I also don't see a field name in the schema?  I am confused here, is this a fixed field, and if so what is its name?

Reviewer 3: This comment applies to line 218, I apparently missed when selecting lines 🤦

Author: The goal here was to say something along the lines of: "any field that's defined for the action object, you can also use here", without duplicating those. But if you think a copy paste instead would make the reading easier, I do that instead, what do you think?

Reviewer 2: Please do not duplicate the fields, that will cause trouble later when fields are added to actions and not here.

Author: Thank you for the input. How do we reword this better so what we mean here is obvious then?

Reviewer 3: @baywet this feels more like a "Patterned Fields" than a "Fixed Fields"?  Also, I'm concerned about embedding an "any field..." in an object with fixed fields.  It invites field name collision problems.  Can we push this down under a field that would hold all of this stuff in its own namespace?

Author: Aaaah the collision aspect is a good callout (from the operations in OpenAPI) I had not thought out. We could in theory create a "definition" field which would effectively be an Overlay Action Object (with all fields optional, and the string interpolation turned on)

What do you think?

(using definition here for lack of better term, to avoid having another thing called "action")

Reviewer 3: Seems like the right general idea, but `definition` has JSON Schema-ish re-use connotations... you could just call it `fields`?

Author: Sure! Always happy to get better name suggestions.

Author: Sure! Always happy to get better name suggestions.

Reviewer: So the goal is that you can re-use something but adjust it for example by setting a more specific description? I see the use case but I think we need to work on how we describe/represent that. I know Ralf doesn't want the fields duplicated but what if we DO later want to add a field there that can't be used here?

Author: Here is my suggestion for this:

- At the reusable action level, we define "fields" instead. And that's just an Action object with two caveats, all fields all optional, any string field may use the string interpolation syntax. Making this change solves two issues: the risk of collision in the future (e.g. what if we want to introduce a description that's for the reusable action itself?) and the "trick" of saying "well you can use any field defined over there" which is odd for sure.
- At the reference level:
  - we *could* completely do away with the reference object together, and move the $ref, parameterValues fields in the Action Object, and add conditional statements. I already said that I'm not in favour of that, because we have precedents that conditionals are hard to understand for humans and difficult to encode using JSON schema. (e.g. all the query parameters and headers conditional fields in OAD)
  - we *could* make the same change that Henry suggested for the reusable actions, move everything in fields. But I don't find that ergonomic when you're working next to action objects. (see examples below)
  - we *could* leave it as is, but it seems to add confusion when reading the specification.
  - we *could* move this definition (without duplicating the fields themselves) to a "pattern fields" section like Henry suggested before. At least it'd make it a bit more obvious this definition is "not a regular field" but that it "brings several fields along".

 ## Examples

I'm going to be using JSON on purpose to outline the difference in ergonomics between options 2 or 3 and 4 :

### keep at the same level

```json
{
  "actions":
   [
     {
       "remove": true,
       "description": "Regular Action object that removes something",
       "target" : "some target"
     },
     {
       "$ref": "#/components/actions/SomeUpdateAction",
       "target": "some target override"
     }
   ]
}
```

### introduce a fields node at the reference level

```json
{
  "actions":
   [
     {
       "remove": true,
       "description": "Regular Action object that removes something",
       "target" : "some target"
     },
     {
       "$ref": "#/components/actions/SomeUpdateAction",
       "fields":
       {
         "target": "some target override"
       }
     }
   ]
}
```

In this example I find the extra node quite verbose and not very ergonomic when you look at a few actions and reference objects mixed together. Thoughts?

Reviewer 2: Why would Overlays have a problem with fields/keywords next to `$ref` when both JSON Schema and OpenAPI allow that?

Author: Thanks everyone, I have made the changes to encapsulate the reusable action fields in 37a0efbf7fd6bc07c91ebd3a08fddddde470b706
Let me know if you have any additional comments or questions.

| <a name="action-template-environment-variables"></a>environmentVariables | [[Action template parameter object](#action-template-parameter-object)] | A list of environment variables to be used during  string literal replacement. Optional. |
| Any field defined in the [action object](#action-object) | mixed | The [string literal replacement syntax](#string-literal-replacement-syntax) MAY be used for any of the fields, and the replacements MUST be evaluated as the template reference is being resolved. |

This object MAY be extended with [Specification Extensions](#specification-extensions).

#### Action template parameter object

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |
| <a name="action-template-parameter-name"></a>name | `string` | **REQUIRED** The name of the parameter, MUST match the key for the string literal replacement expression. |
| <a name="action-template-parameter-default>default | Any | A default value for the parameter to use when no value is provided by the reference. Optional. |

This object MAY be extended with [Specification Extensions](#specification-extensions).

> #### Action reference object

Reviewer 2: ```suggestion
#### Action Reference Object
```

#### Action template reference object

##### Fixed fields

| Field Name | Type | Description |
| ---- | :----: | ---- |
| <a name="action-template-reference-ref"></a>$ref | `string` | **REQUIRED** A valid reference to an action template in the components section, represented as `#/components/actionTemplates/actionTemplateKey`. |
| <a name="action-template-parameter-values>parameterValues | `Map(string, Any)` | A set of values to use for the template parameters, the key MUST match the parameter name. Optional. |

This object MAY be extended with [Specification Extensions](#specification-extensions).

### String literal replacement syntax

The following string replacement syntax MAY be used:

%source.key%

Where **source** is a known source which value MUST ONLY be **env** for environment variables, or **param** for parameters. The key MUST only contain the following characters A-Za-z0-9, contain at least one character and start with a letter.

> #### ABNF notation

Reviewer 2: ```suggestion
#### ABNF Notation
```

```abnf
replacement-string = "%" source "." key "%"
source             = "env" / "param"
key                = ALPHA *( ALPHA / DIGIT )
```

Where ALPHA and digit rules are defined in [[RFC5234]].

#### Environment replacement source

> When the environment replacement source is used, the key MUST match (case-insensitive) an environment variable defined for the process parsing the Overlay document. If the environment variable is not defined, the processor MAY either replace it with an empty string or return an error.

Author: ahhh... PowerShell user here...

| Field Name | Type | Description |
| ---- | :----: | ---- |

> | <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, where that the fragment syntax is JSON Pointer with the pointer prefix restricted to `/components/actions/`. Component action keys in this pointer MUST be encoded according to [[RFC6901], Section 3](https://www.rfc-editor.org/rfc/rfc6901#section-3). |

Reviewer 5: > where that the fragment syntax is JSON Pointer

"where the fragment is a JSON Pointer"
(If that is what you mean?)

Author: ```suggestion
| <a name="reusable-action-reference-ref"></a>$ref | `string` | **REQUIRED** A [same-document](https://www.rfc-editor.org/rfc/rfc3986.html#section-4.4) (or fragment-only) relative URI reference, per RFC3986 §4.4, where the fragment syntax is JSON Pointer with the pointer prefix restricted to `/components/actions/`. Component action keys in this pointer MUST be encoded according to [[RFC6901], Section 3](https://www.rfc-editor.org/rfc/rfc6901#section-3). |
```

Author: Thanks for the callout! fixed.

This example shows how a reusable action can provide shared update content while a reusable action reference supplies the `target` locally. In this case, each reference reuses the same error response definition but supplies a different `target` so the update is applied to different operations.

###### Source Description

```yaml
openapi: 3.2.0
info:
  title: Example API
  version: 1.0.0
paths:
  /items:
    get:
      responses:
        200:
          description: OK
  /some-items:
    delete:
      responses:
        200:
          description: OK
```

###### Overlay

```yaml
overlay: 1.2.0
info:
  title: Use reusable actions to insert error responses
  version: 1.0.0
components:
  actions:
    errorResponse:
      fields:
        update:
          404:
            description: Not Found
            application/json:
              schema:
                type: object
                properties:
                  message:
                    type: string
        description: Adds an error response to the operation
actions:
- $ref: '#/components/actions/errorResponse'
  target: "$.paths['items'].get.responses"
- $ref: '#/components/actions/errorResponse'

>   target: "$.paths['some-items'].delete.responses"

Reviewer 3: Don't `items` and `some-items` need the `/` prefix?  It is part of the property name, if this syntax is doing a full match on property names (I am not very familiar with JSONPath).

Author: ```suggestion
- $ref: '#/components/actions/errorResponse'
  target: "$.paths['/items'].get.responses"
- $ref: '#/components/actions/errorResponse'
  target: "$.paths['/some-items'].delete.responses"
```

Aaaaah very good catch! I can't believe I missed it all this time. Thanks!

```

###### Result Description

```yaml
openapi: 3.2.0
info:
  title: Example API
  version: 1.0.0
paths:
  /items:
    get:
      responses:
        200:
          description: OK
        404:
          description: Not Found
          application/json:
            schema:
              type: object
                properties:

>                   errorMessage:

Reviewer 4: Shouldn't this be 
```suggestion
                  message:
```

                    type: string
  /some-items:
    delete:
      responses:
        200:
          description: OK
        404:
          description: Not Found
          application/json:
            schema:
              type: object
                properties:

>                   deleteErrorMessage:

Reviewer 4: here too?
```suggestion
                  message:
```

          description: OK
```

##### Reusable Action Reference Example - Overrides

This example shows how a reusable action can provide shared update content while a reusable action reference supplies the `target` locally. In this case, each reference reuses the same error response definition but supplies a different `target` so the update is applied to different operations.

###### Source Description

```yaml
openapi: 3.2.0
info:
  title: Example API
  version: 1.0.0
paths:
  /items:
    get:
      responses:
        200:
          description: OK
  /some-items:
    delete:
      responses:
        200:
          description: OK
```

###### Overlay

```yaml
overlay: 1.2.0
info:
  title: Use reusable actions to insert error responses
  version: 1.0.0
components:
  actions:
    errorResponse:
      fields:
        update:
          404:
            description: Not Found

>             application/json:

Reviewer: I think we're missing a `content:` keyword here to make the OpenAPI output correct

Author: thanks, fixed in 031eb9d090cf105d1c86b131c45d562369876958

              schema:
                type: object
                properties:
                  message:
                    type: string

>         description: Adds an error response to the operation

Reviewer: If this is the description for the reusable action (I think it is), then it should be at the same level as `fields` rather than inside it. But if it's at this level perhaps it becomes the action description and this was intentional ... I am not sure how to make this clearer but I did find it confusing.

Author: Thanks for the callout, fixed in 9e4c18af8120e8fcce09f2f8f4bccba92cc65549

actions:
- $ref: '#/components/actions/errorResponse'
  target: "$.paths['/items'].get.responses"
- $ref: '#/components/actions/errorResponse'
  target: "$.paths['/some-items'].delete.responses"
```

###### Result Description

```yaml
openapi: 3.2.0
info:
  title: Example API
  version: 1.0.0
paths:
  /items:
    get:
      responses:
        200:
          description: OK
        404:
          description: Not Found
          application/json:
            schema:
              type: object
                properties:
                  message:
                    type: string
  /some-items:
    delete:
      responses:
        200:
          description: OK
        404:
          description: Not Found
          application/json:
            schema:
              type: object

>                 properties:

Reviewer: Indent level for `properties`: should probably be at the same level as type? (same for the example above as well)

Author: thanks for the callout, fixed in 4f30a3d4da79fe28d85ca39b7a1f7d34ae3eae6a

          description: OK
```

##### Reusable Action Reference Example - Overrides

This example shows how a reusable action can provide shared update content while a reusable action reference overrides one or more action fields locally. In this case, each reference reuses the same error response definition but supplies a different `target` so the update is applied to different operations.

###### Source Description

```yaml
openapi: 3.2.0
info:
  title: Example API
  version: 1.0.0
paths:
  /items:
    get:
      responses:
        200:
          description: OK
  /some-items:
    delete:
      responses:
        200:
          description: OK
```

###### Overlay

```yaml
overlay: 1.2.0
info:
  title: Use reusable actions to insert error responses
  version: 1.0.0
components:
  actions:
    errorResponse:
      update:
        404:
          description: Not Found
          application/json:
            schema:
              type: object
              properties:
                message:
                  type: string

>       target: "$.paths['placeholder'].get.response"

Reviewer: I don't think the target field should be in the reusable action at all. The target/action pairs are the main content of the overlay (I realise the payload is not exactly like that because we published it as-was, but conceptually), and the reusable action should be just that - the action. The targetting is done already and we're just referencing which action to perform.

Author: Also resolving this duplicate feedback in favour of a centralized discussion here https://github.com/OAI/Overlay-Specification/pull/296#discussion_r3073078242

        200:
          description: OK
        404:
          description: Not Found
          application/json:
            schema:
              type: object
                properties:
                  deleteErrorMessage:
                    type: string
```

##### Reusable Action Example - Parameters

This example shows how a reusable action can define parameterized strings and environment-variable placeholders that are resolved separately for each reusable action reference. In this case, each reference supplies different parameter values so the same reusable action can target different operations and produce slightly different response content, while default values are used when a parameter or environment variable value is not otherwise provided.

###### Source Description

```yaml
openapi: 3.2.0
info:
  title: Example API
  version: 1.0.0
paths:
  /items:
    get:
      responses:
        200:
          description: OK
  /some-items:
    delete:
      responses:
        200:
          description: OK
```

###### Overlay

```yaml
overlay: 1.2.0
info:
  title: Use reusable actions to insert error responses
  version: 1.0.0
components:
  actions:
    errorResponse:

>       target: "$.paths['%param.pathItem%'].%param.operation%.responses"

Reviewer: Again, the target doesn't seem to make much sense here, I'd expect the targets to be set as we usually do in Overlays, but to allow referencing an action so it's easily reusable and can be parameterised. I'm really sorry if I missed this in the earlier rounds of review, but I don't think target, or remove, should be here.

Author: I'll go ahead and resolve this one in favour of a centralized discussion on the topic here https://github.com/OAI/Overlay-Specification/pull/296#discussion_r3073078242

      update:
        404:
          description: Not Found
          application/json:
            schema:
              type: object
              properties:
                '%param.propertyName%':
                  type: string
                stageName:
                  type: string
                  const: '%env.stageName%'
      description: Adds an error response to the %param.pathItem% path item %param.operation% operation
      parameters:
      - name: pathItem
      - name: operation
        default: get
      - name: propertyName
        default: errorMessage
      environmentVariables:
      - name: stageName
        default: dev
actions:
- $ref: '#/components/actions/errorResponse'
  parameterValues:
    pathItem: '/items'
- $ref: '#/components/actions/errorResponse'
  parameterValues:
    pathItem: '/some-items'
    operation: delete
    propertyName: deleteErrorMessage
```

> > Note: in this example, no value is set for the process environment variable "stageName".

Reviewer: (I know I'm nitpicking) Do we use note notation like this in our specifications? I really can't remember but let's check it's consistent.

Author: This might be a "muscle memory" reflex from docfx on my end... What would be a better format?

Author: actually I found it somewhere else in the document. updated to `**Note:**`. thanks for the callout.
